lookout/style/typos/utils.py

```diff
@@ -0,0 +1,177 @@
+                               axis=0)
+    flat_column = list(chain.from_iterable(flat_column))
+    result = pandas.DataFrame(flat_values, columns=data.columns)
+    result[new_column] = flat_column
+    return result.infer_objects()
+
+
+def flatten_data(data: pandas.DataFrame, new_column_name="identifier") -> pandas.DataFrame:
+    """
+    Flatten identifiers data in column "token_split"
+    :param data: DataFrame containing column "token_split" with splitted identifiers
+        either as strings or as lists of tokens
+    :param new_column_name: Name of column to put tokens from splits to
+    :return:
+    """
+    apply_function = (lambda x: x) if isinstance(data.token_split.tolist()[0], list) \
+        else (lambda x: str(x).split())
+    return flatten(data, "token_split", new_column_name, apply_function=apply_function)
+
+
+def add_context_info(data: pandas.DataFrame) -> pandas.DataFrame:
+    """
+    Split context of identifier on before and after part
+    :param data: DataFrame. Column "token_split" will be used for
+        creating context info if present
+    :return: Provided data with added columns "before" and "after", containing lists
+        of corresponding contexts tokens
+    """
+    if "before" in data.columns and "after" in data.columns:
+        return data
+
+    tokens = list(data.typo)
+
+    if "token_split" in data.columns:
+        token_split = list(data.token_split)
+        before = []
+        after = []
+
+        for row_number in range(len(data)):
+            if "token_split" in data.columns:
+                split = token_split[row_number]
+                if isinstance(split, str):
+                    split = split.split()
+                index = split.index(tokens[row_number])
+                before.append(split[:index])
+                after.append(split[index + 1:])
+
+    else:
+        before = [[] for _ in range(len(data))]
+        after = [[] for _ in range(len(data))]
+
+    data["before"] = before
+    data["after"] = after
+
+    return data.infer_objects()
+
+
+def rank_candidates(candidates: pandas.DataFrame, pred_proba: list, n_candidates: int = None,
+                    return_all: bool = True):
+    """
+    Rank candidates for typos correction based on correctness probabilities.
+    :param candidates: Dataframe with columns "id", "typo", "candidate"
+        and indexed by range(len(pred_proba))
+    :param pred_proba: Array of probabilities of correctness of every candidate
+    :param n_candidates: Number of most probably correct candidates to return for each typo
+    :param return_all: False to return corrections only for
+        corrected in the first candidate typos
+    :return: Dictionary {id : [[candidate, correct_prob]]}, candidates are sorted
+        by correct_prob in a descending order.
+    """
+    suggestions = {}
+    corrections = []
+    for i in range(len(pred_proba)):
+        index = candidates.loc[i, "id"]
```

Comment:
It's better to make literals `id, candidate, typo, etc` - global constants. In case of refactoring it will require modification in one place only